packages/profiling-browser/src/hubextensions.ts

```diff
@@ -0,0 +1,254 @@
+import { WINDOW } from '@sentry/browser';
+import { getMainCarrier } from '@sentry/core';
+import type { CustomSamplingContext, Hub, Transaction, TransactionContext } from '@sentry/types';
+import { logger, uuid4 } from '@sentry/utils';
+
+import { sendProfile } from './integration';
+import type { JSSelfProfile, JSSelfProfiler, ProcessedJSSelfProfile } from './jsSelfProfiling';
+
+// Max profile duration.
+const MAX_PROFILE_DURATION_MS = 30_000;
+
+// While we experiment, per transaction sampling interval will be more flexible to work with.
+type StartTransaction = (
+  this: Hub,
+  transactionContext: TransactionContext,
+  customSamplingContext?: CustomSamplingContext,
+) => Transaction | undefined;
+
+/**
+ * Check if profiler constructor is available.
+ * @param maybeProfiler
+ */
+function isJSProfilerSupported(maybeProfiler: unknown): maybeProfiler is typeof JSSelfProfiler {
+  return typeof maybeProfiler === 'function';
+}
+
+/**
+ * Wraps startTransaction and stopTransaction with profiling related logic.
+ * startProfiling is called after the call to startTransaction in order to avoid our own code from
+ * being profiled. Because of that same reason, stopProfiling is called before the call to stopTransaction.
+ */
+export function wrapTransactionWithProfiling(
+  transaction: Transaction | undefined,
```

Comment:
I wanted to avoid throwing in case of https://github.com/getsentry/sentry-javascript/issues/4731. I'm not opposed to removing, but if you for example not import @sentry/tracing and end up calling startTransaction then it would look like the profiling package is the cause of errors.

Reply:
Got it. That is reasonable.

We could still make this function only take `Transaction` and change `__PRIVATE__wrapStartTransactionWithProfiling` down below to just return undefined instead of wrapping.

That would allow us to limit a lot of checks in this function. Also adding `| undefined` at a later point in time is not breaking whereas the reverse is, so I would go with that for now.

Reply:
I made this change and I also changed the name of the exported wrap fn to indicate that it is only for the automated instrumentation.